### ISSUE TITLE
Local Development of Postgres DB for Data Model Draft

### ISSUE THREAD

**Issue by zappfyllc** [cityofaustin/atd-data-tech#2326]
_Context_: @zappfyllc has deployed two temporary Heroku applications, one employing GraphiQL and one employing [json-server](https://www.npmjs.com/package/json-server). A flat file in json-server serves as the evolving data model so that queries can be tested with the in-progress schema, specifically to facilitate conversations with Nathan, the architect of the Access DB. This tool will be helpful in the future, however, Hasura allows quick deployment of a Postgres database and automatically writes root queries based on the data model created in the database. Hasura also has a GraphiQL tool implemented in its service. Thus, it may make sense to actually go ahead and deploy Hasura in drafting the data model rather than adjusting the flat file on json-server.

Note: Hasura deployment for Vision Zero Viewer was done through Docker.

UPDATE 8-19-20: [Heroku endpoint for initial Hasura development](https://coa-moped.herokuapp.com/console/data/schema/public)

**Comment by zappfyllc** [comment]
Blocked by developments with dotMaps and completion of ERD/data model; will return to this later

**Comment by zappfyllc** [comment]
Hasura + Heroku w/ Postgres backend will be the infrastructure supporting the initial development of the Data Model by @zappfyllc. Points adjusted for time.

**Comment by zappfyllc** [comment]
Primary Keys Update/Discussion (cc @sergiogcx @mateoclarke @amenity @johnclary):

There are valid arguments for utilizing both UUIDs (universally unique identifiers) and serial keys (most commonly for our purposes, an auto-incrementing integer) as the data type for primary keys in tables (see references below). Based on Moped's feature set, requirements, and imagined future use cases, I am using the following criteria to help determine whether a given table's primary key should be stored in the UUID format or serial format.

- Is the table more immutable than mutable? If a table is likely to be unchanged over time and the list of values is minimal (<10), then I have defaulted the table's primary key as serial (auto-increment). This is because differences across environments (prod, dev, staging, etc.) are highly unlikely and an auto-increment is more dev-friendly and uniqueness matters less.

- Is the table involved in a many-to-many relationship? If a table is involved in a many-to-many relationship where a join table sits between the two main entities involved, consideration should be given to the size of each table and likelihood of querying against the table. In many of these cases, the table that is smaller can get away with having a serial (auto-increment) primary key.

- Is the table a joining table (see above)? If a table sits between two tables in a many-to-many relationship, it may not even need a primary key, as the foreign keys for the table can suffice as unique identifiers for each row. However, a serial (auto-increment) primary key still allows ease of use in terms of viewing the latest data and may provide some value without any real limitation/drawback since the size of serial values is unlikely to have a perceptible impact on performance.

- Is the table's data likely to be distributed across many systems (microservice architecture)? UUIDs are vital for this purpose. Thus, there's an argument to be made that any table that will ultimately be distributed across systems should contain a UUID.

I am very much leaning towards maintaining a UUID as primary key for moped_projects, while leaning towards serials as primary keys for the rest (maybe a handful of situations where this isn't the case, I'm looking into them). There may be some use cases I'm not considering, but I'm curious as to anyone's thoughts on this issue. We can open a new issue if we want to dig deeper.

References:
[Auto Increment Keys vs. UUID] (https://medium.com/@Mareks_082/auto-increment-keys-vs-uuid-a74d81f7476a)
[UUID Performance Concerns] (https://www.percona.com/blog/2019/11/22/uuids-are-popular-but-bad-for-performance-lets-discuss/)
[Does UUID impact writes] (https://stackoverflow.com/questions/34230208/uuid-primary-key-in-postgres-what-insert-performance-impact)

**Comment by johnclary** [comment]
thanks @zappfyllc for the thorough explanation. this makes sense to me and sounds good. agree that UUIDs are overkill in many regards. another approach (maybe) would be to use [MD5 hashses](https://www.postgresqltutorial.com/postgresql-md5/) in places where we are inclined to use UUIDs, with the thinking that we can feel confident the IDs would be globally across all records in the Moped DB. I only suggest this if there were less of a performance hit with MD5's vs UUIDs.

**Comment by johnclary** [comment]
also, it looks like the `uuid-ossp` [extension](https://www.postgresql.org/docs/current/uuid-ossp.html) supports hashes with a smaller footprint. found via [this SO question](https://stackoverflow.com/a/39444532/2103784).

**Comment by zappfyllc** [comment]
Helpful info for building UUID-OSSP if we go that direction: "Historically this module depended on the OSSP UUID library, which accounts for the module's name. While the OSSP UUID library can still be found at http://www.ossp.org/pkg/lib/uuid/, it is not well maintained, and is becoming increasingly difficult to port to newer platforms. uuid-ossp can now be built without the OSSP library on some platforms. On FreeBSD, NetBSD, and some other BSD-derived platforms, suitable UUID creation functions are included in the core libc library. On Linux, OS X, and some other platforms, suitable functions are provided in the libuuid library, which originally came from the e2fsprogs project (though on modern Linux it is considered part of util-linux-ng). When invoking configure, specify --with-uuid=bsd to use the BSD functions, or --with-uuid=e2fs to use e2fsprogs' libuuid, or --with-uuid=ossp to use the OSSP UUID library. More than one of these libraries might be available on a particular machine, so configure does not automatically choose one." [(Source)](https://www.postgresql.org/docs/9.4/uuid-ossp.html)

We could definitely experiment using MD5 hashes (in the Project entity with Project Name, for example) but I'll need to look into whether that actually improves performance anymore than UUID. My one concern with abandoning UUID's altogether throughout the model is potential ETL or ingest down the road with project-level data that might be of interest to other groups both internal and external to COA, plus avoiding any potential issues with distributed services that may be created/utilized (via third-party tools or custom down the road). However, I can't really articulate good examples here.

**Comment by johnclary** [comment]
yuck, then maybe not UUID-OSSP then. i'm not totally sold on the idea that incremental IDs are a blocker to external integrations, but I don't have a lot of experience there. ccing @sergiogcx .

**Comment by sergiogcx** [comment]
If the benchmark is strictly on performance, there isn't all that much difference between integers and uuids in PostgreSQL. I tend to go for keeping things as simple as possible and to avoid unnecessary complexity, and for me it boils down to ease of use and developer experience.

If the decision is between MD5's and UUIDs, I'd stick with UUIDs. If it is between UUIDs and integers, I'd want to know what problems are we trying to solve with UUIDs. If nothing critical, then I'd keep it simple and stick with integers.

**Comment by sergiogcx** [comment]
If you need a technical opinion, here are my two cents on UUIDs:

Pros:
1. They are easy to shard data with
2. Data PKs remain unique when merging in datasets from multiple sources, they can make data integrations easier.
3. It could simplify transactional logic when you need to know your PK before insertion.
4. UUIDs help privacy, it isn't as vulnerable to enumeration attacks (e.g., not as easy to guess if you are user 123 and you are looking for 124)
5. Sequential UUIDs are possible in PostgreSQL (i.e., using a library such as `sequential-uuids`).

Cons:
1. Implementation pain. In other words, uuids are hard to read, harder to update and maintain, adds code complexity to the stack in the front-end as well as back-end, etc. Migrations and refactoring can be painful down the line.
2. Size: In Postgres specifically, a UUID is 4x the size of an INT. (16 bytes for a uuid vs. 4 for an int), but it can be 9x larger in memory in any system that deals with them as a string (36 bytes).
3. UUIDs don’t sort as fast as numbers because they rely on collation rules, you need other mechanisms for sorting quickly.
4. **Table indexes can exponentially bloat**, it gets worse if you have many tables referencing others with uuids, each table needs to have its own uuid index. Imagine having intermediary tables (ie. many-to-many relationships) with millions of records for things like addresses, phone numbers, users, etc. Bloat is not as prevalent for denormalized tables though.
5. Higher memory use, when joining tables and sorting the index data needs to live in disk and in RAM memory. Joins can become noticeably slower here, possibly causing bottlenecks in the stack.
6. More database maintenance needed, UUIDs are random and cannot be clustered, they cause more memory fragmentation, which means we need to use things like Autovaccum more often.
7. In PostgreSQL, WAL writes to read-replicas are slower in large datasets. This means the wait time "could" be longer to see the changes propagated to read replicas from its master node.

My main concerns with UUIDs is added dev complexity and index bloat, looking at the database we do have quite a few tables that derive from the projects table, meaning we need lots of relationships. If we are expecting lots a lot of data, we are likely to experience the latter. We can mitigate by running a larger database instance, bloat is a problem for smaller/medium instances.

**Comment by zappfyllc** [comment]
Thanks @sergiogcx, that is a helpful technical breakdown and aligns with what I've read. It seems you and @johnclary are on the same page and so I feel more comfortable closing the UUID discussion and deciding on integers, with the understanding that #2 and #4 in your pros list are my biggest concerns but can likely be solved by 1) centralizing writes to project-related data in Moped and 2) ensuring user authentication for any URLs that expose IDs.

**Comment by sergiogcx** [comment]
@zappfyllc Thank you for working on the database JD, its definitely one of the hardest parts of any project.

I think the integers will work fine, but as an additional proposal, it would not be a bad idea to keep the UUID in the projects table so we can use it for sharing, what do you think? The UUID would still be a unique-index but it would not be a primary key, so that way any possible index bloat is limited to one table and we can use the UUID whenever it is convenient (e.g., sharing the project as an external link, etc.)

If it is too much fuss, then let's just stuck with ints.

**Comment by zappfyllc** [comment]
@sergiogcx 100% agreed, not a fuss at all!

**Comment by zappfyllc** [comment]
Moving this to Sprint 32 for finishing the following:
- Reducing database scope (remove tables that won't be included in first few releases)
- Building queries to test out working relationships and identifying missing relationships
- Time-pending: "Summary" entities, 1:1 project entity destructuring proposal: I need to set up moped_proj_timeline and a few other of these tables that are 1:1 with project and how they pull in information from related projects

Wherever we end up, we'll close this out in Sprint 32 and start logging database changes in accordance with #2410 from then on out

**Comment by zappfyllc** [comment]
@sergiogcx As of 9/15, data model was updated again after your ERD generation that visually displayed some of the gaps I missed. I removed a couple tables and added relationships. Just including you here in case it impacts your migration next week in any way!

**Comment by zappfyllc** [comment]
Closing this after migration of Postgres DB to AWS RDS.
